DOCS-9680 Commerce Cloud Shop Connector (#1184)
mulesoft/docs-connectors: commerce-cloud-shop-api/1.0/modules/ROOT/pages/shop-api-connector-examples.adoc

= Commerce Cloud B2C Shop Connector Example - Mule 4

This example for Anypoint Connector for Salesforce Commerce Cloud B2C Shop (B2C Shop Connector) example shows you how to:

* Use the *Authorize Customer* operation to get a Shopper JWT token. 

> * Use the *Get Customer* operation to view the customer's details.  

Reviewer: Capitalize operation names but do not boldface in this context, because we are not instructing the audience in the context of viewing the UI label.

> This example uses the B2C Shop Connector *Customer Auth Connection Provider* configuration.

Reviewer: ditto

== Authorize Customer

> Configure the *HTTP Listener* source and *Authorize Customer* operation to retrieve the Shopper JWT token from Postman:

Reviewer: ditto

. In Studio, select *File > New > Mule Project*.
. Enter a name for your Mule project and click *Finish*.
. In the *Mule Palette* view, select *HTTP > Listener*.
. Drag *Listener* to the Studio canvas.
. On the *Listener* configuration screen, optionally change the value of the *Display Name* field.

> . Specify `/customer/shopper-customers/{version}/organizations/{organizationId}/customers/actions/login` for the *Path* field.

Reviewer: ```suggestion
. Specify `/customer/shopper-customers/{version}/organizations/{organizationId}/customers/actions/login` for the *Path* field:
```

+
image::shop-connector-authorize-customer-http-listener-config.jpg[HTTP Listener General properties tab]
+

> . Click the plus sign (*+*) next to the *Connector configuration* field to configure a global element that can be used by all instances of HTTP *Listener* in the app.

Reviewer: Do you mean to refer to all instances of the UI label *Listener*?
If not, use plain text font.

. On the *General* tab, specify `8081` for *Port*.

> . In the *Mule Palette* view, select the *Authorize Customer* operation and drag it on to the flow.

Reviewer: ```suggestion
. In the *Mule Palette* view, select the *Authorize Customer* operation and drag it on to the flow:
```

+
image::shop-connector-authorize-customer.png[Mule flow with HTTP Listener as the source and the Authorize Customer operation]
+

> . Obtain the required parameters for the *Authorize Customer* operation from Postman:

Reviewer: ditto: are we referring to the operation in general or the UI label?

+
image::shop-connector-authorize-customer-body.jpg[Authorize Customer operation General tab]
+

> . Select the *Commerce Cloud B2C Shop Connector Shopper Token Config* from Commerce Cloud B2C Shop Connector and configure it with the required values or add a new configuration by clicking the green plus (*+*) symbol.

Reviewer: ```suggestion
. Select *Commerce Cloud B2C Shop Connector Shopper Token Config* from Commerce Cloud B2C Shop Connector and configure it with the required values, or add a new configuration by clicking the green plus (*+*) symbol:
```
Is Commerce Cloud B2C Shop Connector in this context a UI label? If so, bold.

+
image::shop-connector-shopper-token-config.jpg[Commerce Cloud B2C Shop Connector Token Config General properties tab]
+
. After entering all the required parameters, save the Mule project.
. Run the project as a Mule application by right-clicking the project name in *Package Explorer* and selecting *Run As > Mule Application*.
. Navigate to `+http://localhost:8081/customer/shopper-customers/{version}/organizations/{organizationId}/customers/actions/login+`. 

> . Open Postman, and check the response. +

Reviewer: ```suggestion
. Open Postman and check the response, noting the JWT Token in the response header. +
```

> You see the JWT Token in the response header. 

Reviewer: omit per above

> Provide this JWT Token from Postman as the authorization header parameter to perform operations.

Reviewer: Do you consider this a step in the task Authorize Customer?
If so, format as such; if not, consider rewording to something less immediately imperative?

[[xml-authorize-customer]]

> === XML for Authorize Customer

Reviewer: ```suggestion
=== XML for Authorize Customer Operation?
```
See next topic title as well.

> Click the *Configuration XML* tab at the base of the canvas and copy and paste the this XML code into the Studio XML editor to quickly load the flow for this example into your Mule app:

Reviewer: ```suggestion
Click the *Configuration XML* tab at the base of the canvas and copy and paste the following XML code into the Studio XML editor to quickly load the flow for this example into your Mule app:
```

		<http:listener
			doc:name="8081/customer/shopper-customers/{version}/organizations/{organizationId}/customers/actions/login"
			config-ref="HTTP_Listener_config"
			path="/customer/shopper-customers/{version}/organizations/{organizationId}/customers/actions/login"
			allowedMethods="POST">
			<http:response>
				<http:headers><![CDATA[#[output application/java
---
{
"Authorization" : message.attributes.headers.Authorization
}]]]></http:headers>

			</http:response>
			<http:error-response statusCode="#[error.errorMessage.attributes.statusCode]">
				<http:body><![CDATA[#[output text/json --- error.errorMessage.payload]]]></http:body>

			</http:error-response>
		</http:listener>
		<commerce-cloud-shopper-api:create-customer-shopper-customers-organizations-customers-actions-login-by-version-organization-id
			doc:name="Authorize Customer"
			version="#[attributes.uriParams.version]"
			organizationId="#[attributes.uriParams.organizationId]" clientId="#[attributes.queryParams.clientId]"
			siteId="#[attributes.queryParams.siteId]" authorization="#[attributes.headers.Authorization]" config-ref="Commerce_Cloud_B2C_Shop_Connector_Shopper_token"/>

	</flow>
</mule>
----

> == Get Customer

Reviewer: ```suggestion
== Get Customer Operation?
```

> Configure the *HTTP Listener* source and *Get Customer* operation to retrieve and view the customer's details:

Reviewer: ditto re. formatting general object vs. instruction to manipulate UI control indicated by label

. Create a new Mule application in Anypoint Studio (Studio), configuring port 8081 as HTTP Listener and the path as `+/customer/shopper-customers/{version}/organizations/{organizationId}/customers/{customerId}+`:
+
image::shop-connector-http-listener-config.jpg[HTTP Listener General properties tab]
+

> . In the *Mule Palette* view, select the *Get Customer* operation and drag it on to the flow:

Reviewer: Is the audience selecting one of many GC operations, or is the audience selecting the item labeled *Get Customer"?

+
image::shop-connector-get-customer-flow.png[Mule flow with HTTP Listener as the source and the Get Customer operation]
+

> . Obtain the required parameters for *Get Customer* operation from Postman.

Reviewer: ```suggestion
. Obtain the required parameters for the *Get Customer* operation from Postman:
```
ditto: the operation by that name, or the UI label?

+
image::shop-connector-get-customer-body.jpg[Get Customer General properties tab with values for Version, Organization ID, Customer ID, and Site ID]
+

> . Select the *Commerce Cloud B2C Shop Connector Config* from Commerce Cloud B2C Shop Connector and configure it with the required values or add a new configuration by clicking the green plus (*+*) symbol.

Reviewer: Same questions as at Line 100

Reviewer: Also, end with colon.

+
image::shop-api/shop-connector-example/shop-connector-customer-auth-config.jpg[Commerce Cloud B2C Shop Connector Config General properties tab]
+

> In the above image the authentication token from *Authorize Customer* is used as the *Authorization* parameter value.

Reviewer: ```suggestion
The authentication token from *Authorize Customer* shown here is used as the *Authorization* parameter value.
```

+
. After entering the all required parameters, save the Mule project.
+
. Run the project as a Mule application by right-clicking the project name in *Package Explorer* and selecting *Run As > Mule Application*.
+
. Navigate to `+http://localhost:8081/customer/shopper-customers/{version}/organizations/{organizationId}/customers/{customerId}+`. 

> . Open Postman, pass `xref:Authorize Customer<Shopper JWT Token>` from the header and then check the response.

Reviewer: We need a different verb than "pass" here..."supply"?

You see the customer details in the output.

> 

Reviewer: BREAK HERE to enable writer to get a start.

Reviewer: RESTART

[[xml-get-customer]]

> === XML for Get Customer

Reviewer: Operation?

> Click the *Configuration XML* tab at the base of the canvas and copy and paste the this XML code into the Studio XML editor to quickly load the flow for this example into your Mule app:

Reviewer: ```suggestion
Click the *Configuration XML* tab at the base of the canvas and copy and paste the following XML code into the Studio XML editor to quickly load the flow for this example into your Mule app:
```
